### ISSUE TITLE
Bucket_no_upscaling would be a good feature

### ISSUE THREAD

**Issue by Deathawaits4** [tdrussell/diffusion-pipe#262]
I know there has been an issue on this allready:

> It doesn't support disabling upscaling. Images are resized (up or down) to match the resolutions you configured. And probably this behavior will never change. How would it know exactly what size bucket to make a lower res image, where you want to disable upscaling, if you don't explicitly specify that resolution?
> 
> I think you can achieve what you want by just separating out the dataset into different folders for different resolutions. I.e. all ~512 images are in their own folder, and you have a dataset config for just that. Then you can do your phases of training by swapping out the dataset config, and using the init_from_existing option.

but i disagree with this statement and it was only somewhat true for sdxl at times. 

Upscaling images is BAD, very bad especially if you train in multi resolution. And resolutions inbetween 512-1024. Flux has been trained on immense amount of data divisible by 32, it didn't explicitly use 1024, 768, 512 resolutions. bucket_no_upscale would only need to calculate the nearest aspect ratio to one that is divisible by 32 and this would work great for most t2i models. Adding this as an optional feature would be great and save most people hours of work filtering their dataset into buckets. I hope you can reconsider this stance as it has proven to be invaluable in all other trainers available.

**Comment by tdrussell** [comment]
This can be done, but there's at least 2 ways to do it and I'm not sure which would be best.

1. Find nearest size divisible by 32 (rounding down), that's the bucket for the image. The downside of this is that you likely end up with a lot of buckets, and if global batch size is >1, a bunch of images could be dropped because it has to round each bucket to a multiple of the global batch size.
2. Have a bucket_no_upscale option change how multi resolution training works. An image would be assigned to only one of the resolutions, not all of them, such that upscaling never happens. I.e. find the largest resolution that's still smaller than the image. If the image is smaller than the smallest resolution, it would be dropped I guess?

**Comment by Deathawaits4** [comment]
> This can be done, but there's at least 2 ways to do it and I'm not sure which would be best.
> 
> 1. Find nearest size divisible by 32 (rounding down), that's the bucket for the image. The downside of this is that you likely end up with a lot of buckets, and if global batch size is >1, a bunch of images could be dropped because it has to round each bucket to a multiple of the global batch size.
> 2. Have a bucket_no_upscale option change how multi resolution training works. An image would be assigned to only one of the resolutions, not all of them, such that upscaling never happens. I.e. find the largest resolution that's still smaller than the image. If the image is smaller than the smallest resolution, it would be dropped I guess?

Well my idea is either:

Kohya implementation: It duplicates images in a bucket until it fills the batch size, so if you have batch 32 but only 16 images in a bucket, all of them get duplicated. This could cause a slight bias, but in a large dataset this doesn't matter much.

ai-toolkit: Basically instead of training a full batch, the batch size is variable. So images that only have 16 images within a bucket, get trained with a batch size of 16 instead of the 32. This is the better solution IMO, you'd have some stronger outliers that might contribute more since they aren't averaged. 

For situations like these it is also recommended to reduce the batch size and up gradient accumulation because it would average between all resolutions and give the model strong generalization between resolutions. 

Basically if the user knows what he is doing, it can be very powerfull. If people just use batch sizes without knowing what they actually do, it will at worst give the same results.

**Comment by kekm8-cpu** [comment]
> For scaling:
> 
> if you train on multiple resolutions and set "maximum resolution" and your dataset consists of everything between 512 and 4k images:
> 
> Images smaller than the max resolution will not be scaled, they will only be cropped into a 32 divisible bucket while keeping the aspect as close as possible to the original.
> 
> Images larger than 1024 will first get scaled down to the neares aspect ratio matching 1024x1024 pixels and then get cropped to a 32 divisible bucket.

Going to second this. This would be huge.
